[Hardware][AMD][CI/Build][Doc] Upgrade to ROCm 6.1, Dockerfile improvements, test fixes

mawong-amd opened the pull request:
This PR does the following:

1. Upgrade the default base image to ROCm 6.1.2 with the correct fixes applied, and the default PyTorch version to 2.4.0, bringing it in line with CUDA's update to 2.3.0.
2. Refactor to enable multi-stage builds in Dockerfile.rocm: flash-attn and Triton are now built in separate stages as wheels which are then installed inside the final image. vLLM will eventually also be built in a similar manner once AMD CI is compatible. This improves container build time since component stages can be built in parallel.
3. Restore the use of `PYTORCH_ROCM_ARCH` as the first point of reference for what ARCHes to build for. This has not been working correctly since the transition to a CMake based system in https://github.com/vllm-project/vllm/pull/2830, resulting in issues such as that mentioned in https://github.com/vllm-project/vllm/pull/5641.
4. Improve miscellaneous bits of Dockerfile logic: for instance, better detection of ROCm version (as opposed to matching `BASE_IMAGE` exactly), and use of `setup.py develop` instead of `setup.py install` which sidesteps the module-shadowing issues that resulted in the need to copy over the vLLM C++ libraries.
5. Other small miscellaneous documentation fixes.
6. Fix a bunch of tests that were broken with upstream changes.

This PR contains the fixes in https://github.com/vllm-project/vllm/pull/5641 in a different form and can be merged after or without it.

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

mawong-amd committed:
Add ROCm tuned num_threads and partition_size for pagedattn

mawong-amd committed:
Use PYTORCH_ROCM_ARCH env variable to set target list

mawong-amd committed:
Upgrade Dockerfile to ROCm6.1 properly and improve logic, build FA and Triton in separate stages

WoosukKwon commented:
@mawong-amd Thanks for the PR! This is wonderful!

@hongxiayang How would like to merge these two PRs?

mawong-amd committed:
ROCm 6.1.2, reorder dependency installation for vLLM concurrent build

mawong-amd committed:
Standardize use of PyTorch 2.3.0 on ROCm 6.x

mawong-amd committed:
Ensure use of docker buildkit for AMD container

mawong-amd committed:
Use nightly torch for ROCm 6.1, relax req to torch version >= 2.3.0

hongxiayang commented on cmake/utils.cmake:
```diff
@@ -147,16 +147,23 @@ macro(override_gpu_arches GPU_ARCHES GPU_LANG GPU_SUPPORTED_ARCHES)
+    # "rocm_agent_enumerator" in "enable_language(HIP)"
+    # (in file Modules/CMakeDetermineHIPCompiler.cmake)
+    #
+    if(DEFINED ENV{PYTORCH_ROCM_ARCH})
```
Reusing PYTORCH_ROCM_ARCH environment variable in VLLM environment exposes several potential problems: (1) this environment variable was used by pytorch on rocm when building its rocm backend (https://github.com/pytorch/pytorch/blob/0707811286d1846209676435f4f86f2b4b3d1a17/cmake/Dependencies.cmake#L1075). Reuse this will hide this information, and may potentially make the debugging more difficult. (2) Also, without intersecting with pytorch ROCm targets, users can build any vllm-unsupported rocm target. For example, if a certain user posts a community issue on vllm, and that user used this env to build vllm for gfx900, we can imagine the debugging steps needed for supporting such kind of scenarios.

mawong-amd replied:
I'm not really sure if this is a concern:
1. This just restores the behavior to what it was prior to https://github.com/vllm-project/vllm/pull/2830
2. We also now install `torch` wheels, so `PYTORCH_ROCM_ARCH` in the base image is decoupled from whatever "Pytorch used when building its ROCm backend"
3. The debugging info provided by this environment variable seems to be very limited.

hongxiayang replied:
Yes, I am concerned about the name clash. Can you change it to VLLM_ROCM_SUPPORTED_ARCHS?
Though the dockerfile in this PR installed wheels, the wheels was built using the same environment variable and same logic.

mawong-amd replied:
It was working fine prior to https://github.com/vllm-project/vllm/pull/2830 and the workflow people are used to is to specify `PYTORCH_ROCM_ARCH` to decide the architectures to build vLLM for: this has been the case at ROCm/vllm and here, just that it hasn't been working here in a while.. I haven't yet seen a good reason to change this workflow. If this change is made, it should be made in a separate PR with proper discussion of the change.

mawong-amd committed:
Fix up rebase errors, upgrade to torch 2.4.0, install amdsmi

mawong-amd committed:
Fix test_cuda_device_count_stateless on ROCm

mawong-amd committed:
Use OpenAI triton instead of ROCm

mawong-amd committed:
Revert ray workdir init from https://github.com/vllm-project/vllm/pull/5100

mawong-amd committed:
Use 240612 torch wheels, prevent runtime interference

mawong-amd commented:
@WoosukKwon CI is looking all green now  other than Documentation, which has been broken for a while, and the AMD Vision Language Models test, which seems to be due to a slow model download hitting the 90 min timeout (machine issue, seems to have originated sometime in the last).

I've also fixed a bunch of tests that were broken due to:
1. https://github.com/vllm-project/vllm/pull/5100 initializing working directory for ray workers. Reverting this doesn't seem to have affected the tests.
2. Addition of a test that uses `CUDA_VISIBLE_DEVICES`. Fixing this requires updating `torch` to 2.4.0 and installing `amdsmi` for stateless detection of devices, along with using `HIP_VISIBLE_DEVICES`.

Can we land this?

WoosukKwon commented on vllm/utils.py:
```diff
@@ -746,7 +746,8 @@ def _cuda_device_count_stateless(
     # bypass _device_count_nvml() if rocm (not supported)
-    nvml_count = -1 if torch.version.hip else torch.cuda._device_count_nvml()
+    nvml_count = torch.cuda._device_count_amdsmi() if (
+        torch.version.hip) else torch.cuda._device_count_nvml()
```
can we use `is_hip` (in `utils.py`) instead?

mawong-amd replied:
We could, I'm following what I thought was a convention for this file specifically: the `find_nccl_library` function also uses bare `torch.version.hip` and `torch.version.cuda`

WoosukKwon commented on CMakeLists.txt:
```diff
@@ -32,8 +32,7 @@ set(HIP_SUPPORTED_ARCHS "gfx906;gfx908;gfx90a;gfx940;gfx941;gfx942;gfx1030;gfx11
 set(TORCH_SUPPORTED_VERSION_CUDA "2.3.0")
-set(TORCH_SUPPORTED_VERSION_ROCM_5X "2.0.1")
-set(TORCH_SUPPORTED_VERSION_ROCM_6X "2.1.1")
+set(TORCH_SUPPORTED_VERSION_ROCM "2.4.0")
```
QQ: How stable is PyTorch+ROCm 2.4.0?

mawong-amd replied:
We've been using various 2.4.0 pins internally for a while and so far CI has not flagged any issues arising from PyTorch 2.4.0. I've pinned the specific PyTorch 2.4.0 wheel that we install to a known good version. We need PyTorch 2.4.0 because this PR https://github.com/vllm-project/vllm/pull/5473 introduced stateless device counts which are supported on ROCm only from 2.4.0 onwards.

WoosukKwon commented on rocm_patch/libamdhip64.so.6:
Can we avoid adding this binary to the repo? We'd like to keep the repo size as small as possible, and also this will cause huge git diffs whenever we update the file.

WoosukKwon replied:
If this is a general bug fix (or extension) for ROCm 6.1, could you please push it in somewhere else (like in the ROCm 6.1 docker image) and let vLLM download it?

mawong-amd replied:
We're not at all likely to update this file, which brings in improvements to HIP graphs from ROCm 6.2 that we want earlier in vLLM. The risk of using special Docker images is that users will not always know to use these images, as they naturally gravitate towards official images.

How does hosting this on our ROCm/vllm fork and downloading it when building the image instead sound?

WoosukKwon replied:
Sounds good! Thanks for the solution!

WoosukKwon commented on vllm/envs.py:
```diff
@@ -11,6 +11,7 @@
     VLLM_USE_TRITON_FLASH_ATTN: bool = False
     LOCAL_RANK: int = 0
     CUDA_VISIBLE_DEVICES: Optional[str] = None
+    HIP_VISIBLE_DEVICES: Optional[str] = None
```
What about only exposing `CUDA_VISIBLE_DEVICES` to the vLLM users and we internally convert it to `HIP_VISIBLE_DEVICES` when using the AMD GPUs?

mawong-amd replied:
I've done that for these tests for now, we can revisit how we use `CUDA_VISIBLE_DEVICES` generally in a follow-up PR. Copying over any changes to `CUDA_VISIBLE_DEVICES` to `HIP_VISIBLE_DEVICES` is not portable in general. Probably an abstraction layer that either routes device changes to `HIP_VISIBLE_DEVICES` or `CUDA_VISIBLE_DEVICES` is best.

mawong-amd replied:
In general we cannot expect ROCm to recognize `CUDA_VISIBLE_DEVICES`. Unfortunately we currently don't have a unified interface for setting `CUDA_VISIBLE_DEVICES`: each use case (Ray executor, MP executor, custom all reduce, tests, etc) does it in its own way, so there will need to be quite a bit of refactoring, and even then, I'm not sure if there's a better solution than just always making sure that changes to `CUDA_VISIBLE_DEVICES` <=> changes to `HIP_VISIBLE_DEVICES`.

WoosukKwon commented on tests/async_engine/test_openapi_server_ray.py:
```diff
@@ -4,15 +4,15 @@
 @pytest.fixture(scope="module")
 def ray_ctx():
-    ray.init(runtime_env={"working_dir": VLLM_PATH})
+    ray.init()
```
I actually don't understand the purpose of this change. Could you elaborate more on this?

mawong-amd replied:
Two issues: 
1. AMD's container has the entire vLLM development directory under `VLLM_PATH`, while Ray has a 500 MiB package limit for runtime environments. So, there's the risk (especially when running tests locally, when the development directory is further bloated) that we hit the limit and Ray initialization fails.
2. Setting the working directory to `VLLM_PATH` causes `vllm._C ModuleNotFound` errors because we end up shadowing the `vllm` module, which can be installed anywhere (typically the `pythonver/site-packages` folder) with a "local" version in `${VLLM_PATH}/vllm". This is an unavoidable feature of Python: when importing module `foo` the module finder goes through each entry in `sys.path` and looks for either `foo.py` or `foo/__init__.py`. The working directory is typically the first entry in `sys.path`, so `${VLLM_PATH}/vllm/__init__.py` is found before the actual vLLM installation say in `python3.9/site-packages/vllm`.

More importantly, getting rid of this change doesn't seem to affect tests: I'm curious as to why it was added in the first place, but it seems to be unnecessary.

mawong-amd replied:
For 1. you can see an example of the problem [here in the AMD: Entrypoints test](https://buildkite.com/vllm/ci-aws/builds/2175#0190391d-fa82-4ec5-99a1-bdabd605da5f):

```
ERROR entrypoints/test_openai_embedding.py::test_single_embedding[intfloat/e5-mistral-7b-instruct] - ray.exceptions.RuntimeEnvSetupError: Failed to set up runtime environment.
--
  | 2024-06-21 00:32:49 CDT | Failed to upload working_dir /vllm-workspace to the Ray cluster: Package size (594.65MiB) exceeds the maximum size of 500.00MiB. You can exclude large files using the 'excludes' option to the runtime_env or provide a remote URI of a zip file using protocols such as 's3://', 'https://' and so on, refer to https://docs.ray.io/en/latest/ray-core/handling-dependencies.html#api-reference.
  | 2024-06-21 00:32:49 CDT | ERROR entrypoints/test_openai_embedding.py::test_batch_embedding[intfloat/e5-mistral-7b-instruct] - ray.exceptions.RuntimeEnvSetupError: Failed to set up runtime environment.
  | 2024-06-21 00:32:49 CDT | Failed to upload working_dir /vllm-workspace to the Ray cluster: Package size (594.65MiB) exceeds the maximum size of 500.00MiB. You can exclude large files using the 'excludes' option to the runtime_env or provide a remote URI of a zip file using protocols such as 's3://', 'https://' and so on, refer to https://docs.ray.io/en/latest/ray-core/handling-dependencies.html#api-reference.
```

WoosukKwon replied:
@DarkLight1337 It seems `runtime_env={"working_dir": VLLM_PATH}` was introduced in your PR #5100. Could you confirm that this change is safe and explain the reason if not?

WoosukKwon commented on Dockerfile.rocm:
```diff
@@ -39,79 +41,150 @@ RUN apt-get update && apt-get install -y \
 
-RUN python3 -m pip install --upgrade pip
-RUN python3 -m pip install --no-cache-dir fastapi ninja tokenizers pandas
+RUN pip install --upgrade pip
+RUN pip install --no-cache-dir fastapi ninja tokenizers pandas
```
QQ: Why do we have to install those dependencies separately? Aren't they installed while installing `torch` or installing `requirements-common.txt`?

mawong-amd replied:
They were in the previous Dockerfile.rocm (current upstream Dockerfile.rocm) but no longer necessary, and have been removed.

WoosukKwon commented on Dockerfile.rocm:
```diff
@@ -39,79 +41,150 @@ RUN apt-get update && apt-get install -y \
 
-WORKDIR /vllm-workspace
+
+### AMD-SMI build stage
+FROM base AS build_amdsmi
+RUN cd /opt/rocm/share/amd_smi \
+    && pip wheel . --wheel-dir=/install
```
nit: Maybe move this before FA and Triton since it's required as opposed to the other being optional?

mawong-amd replied:
Done, and same for the install stage too.

WoosukKwon commented on Dockerfile.rocm:
```diff
@@ -1,34 +1,36 @@
 ARG FA_GFX_ARCHS="gfx90a;gfx942"
-RUN echo "FA_GFX_ARCHS is $FA_GFX_ARCHS"
-
 ARG FA_BRANCH="ae7928c"
```
Just curious: Do we have to update the CK-based FA? IIRC, we've been using this version for a while.

mawong-amd replied:
We'll look into potentially doing the CK FA update in another PR. We do have an improved CK for MI300 [here](https://github.com/ROCm/flash-attention/tree/ck_tile), but it's not quite yet ready for prime time (ROCm detection doesn't work correctly when building, supports MI300 only, etc.).

WoosukKwon commented on Dockerfile.rocm:
```diff
@@ -39,79 +41,150 @@ RUN apt-get update && apt-get install -y \
     && cd triton/python \
-    && pip3 install . \
-    && cd ../..; \
+    && python3 setup.py bdist_wheel --dist-dir=/install; \
```
nit: Do we want to pin a specific version?

WoosukKwon commented on Dockerfile.rocm:
```diff
@@ -39,79 +41,150 @@ RUN apt-get update && apt-get install -y \
 # Workaround for ray >= 2.10.0
 ENV RAY_EXPERIMENTAL_NOSET_ROCR_VISIBLE_DEVICES=1
+# Silences the HF Tokenizers warning
+ENV TOKENIZERS_PARALLELISM=false
```
QQ: can this have any performance impact?

mawong-amd replied:
It doesn't have any performance impact over the status quo, because currently HF Tokenizers disables parallelism automatically, but with a warning. This env var is just to silence the warning.

WoosukKwon commented on Dockerfile.rocm:
```diff
@@ -39,79 +41,150 @@ RUN apt-get update && apt-get install -y \
+    && python3 setup.py clean --all \
+    && python3 setup.py develop
+
+# Copy triton wheel(s) if any
+RUN --mount=type=bind,from=build_triton,src=/install,target=/install \
+    mkdir -p libs \
+    && if ls /install/*.whl; then \
+        cp /install/*.whl libs \
+        # Preemptively uninstall to avoid same-version no-installs
+        && pip uninstall -y triton; fi
 
+# Copy flash-attn wheel(s) if any
+RUN --mount=type=bind,from=build_fa,src=/install,target=/install \
+    mkdir -p libs \
+    && if ls /install/*.whl; then \
+        cp /install/*.whl libs \
+        # Preemptively uninstall to avoid same-version no-installs
+        && pip uninstall -y flash-attn; fi
+
+# Copy amdsmi wheel(s) if any
+RUN --mount=type=bind,from=build_amdsmi,src=/install,target=/install \
+    mkdir -p libs \
+    && if ls /install/*.whl; then \
+        cp /install/*.whl libs \
+        # Preemptively uninstall to avoid same-version no-installs
+        && pip uninstall -y amdsmi; fi
+
+# Install any dependencies that were built
+RUN --mount=type=cache,target=/root/.cache/pip \
+    if ls libs/*.whl; then \
+    pip install libs/*.whl; fi
```
What are these lines for?

mawong-amd replied:
Currently we build vLLM as follows:
`triton`, `flash-attn`, and `amdsmi` each get a build stage, whose only step is to build the respective wheels. `vLLM` also gets built concurrently, but in the `final` stage, which ultimately generates the final image.

The `final` stage has to take in the results of the `triton`, `flash-attn`, and `amdsmi` stages, i.e., their wheels, so they can be installed. These steps do exactly that: they 1) copy over the wheels into the final image for reference, 2) uninstall their respective module to ensure the respective wheel gets installed (`pip` may refuse if not due to what it believes is installing a same/lower version module), and 3) install all the wheels together.

mawong-amd replied:
Also worth noting that these steps have to be done after the vLLM build, otherwise the vLLM build will not be done concurrently (because in theory it could depend on the previous steps).

WoosukKwon replied:
Oh I see. I didn't recognize that we only built the wheels but not actually installed them. BTW, can we change the comments from `Copy flash-attn wheel(s) if any` to `Copy the flash-attn wheel if it's built` or something like that?

mawong-amd replied:
Oops forgot to comment earlier but these have been updated, and the build stages have an extra comment to specify that they are building wheels.

WoosukKwon reviewed (commented):
@mawong-amd Thanks for the PR and sorry for the delays in the review. This is big!!

I'm generally good for the changes, but have some questions. Please check out my comments.

Also, I think we can merge this PR faster if you can split the PR into smaller ones. While the PR size is not big, I feel it touches quite many parts of the code. I'd like to integrate some of the changes first, and discuss the rest of the changes more deeply. For example, you could split the PR into 4 smaller PRs:
1. Update PagedAttention hyperparameters
2. Fix the docs
3. Refactor the docker file
4. Update the docker file with torch 2.4.0 & Introduce `HIP_VISIBLE_DEVICES`

mawong-amd commented:
> @mawong-amd Thanks for the PR and sorry for the delays in the review. This is big!!
> 
> I'm generally good for the changes, but have some questions. Please check out my comments.
> 
> Also, I think we can merge this PR faster if you can split the PR into smaller ones. While the PR size is not big, I feel it touches quite many parts of the code. I'd like to integrate some of the changes first, and discuss the rest of the changes more deeply. For example, you could split the PR into 4 smaller PRs:
> 
> 1. Update PagedAttention hyperparameters
> 2. Fix the docs
> 3. Refactor the docker file
> 4. Update the docker file with torch 2.4.0 & Introduce `HIP_VISIBLE_DEVICES`

Hi @WoosukKwon, thanks for your review! The top priority for us is updating the container to ROCm 6.1 (i.e. item 3), so I'll work on that PR first. I think it makes sense to combine 2, 3, and 4. For 2, the doc changes mostly relate to how we build the container. And we have to do 4 before or at the same time as 3, otherwise we will have failing tests while upgrading to 6.1. Timing is somewhat important here: the longer it takes to do the 6.1 upgrade, the more upstream tests will break on AMD, and the more such tests we have to fix before we can do the upgrade.

mawong-amd committed:
Use CUDA_VISIBLE_DEVICES for ROCm tests

mawong-amd committed:
Reviewer comments on Dockerfile

mawong-amd committed:
Unify setting CUDA_VISIBLE_DEVICES and carrying it over to HIP_VISIBLE_DEVICES on ROCm

hongxiayang commented:
> > 1. Update PagedAttention hyperparameters
> 
I agree about pulling out the change for "NUM_THREADS = 1024" to a different pr. As we discussed previously, do you know how  this change impact non-mi300x uses?

hongxiayang commented on Dockerfile.rocm:
```diff
@@ -39,79 +40,155 @@ RUN apt-get update && apt-get install -y \
+        *"rocm-5.7"*) \
+            pip uninstall -y torch \
+            && pip install --no-cache-dir --pre torch==2.4.0.dev20240612 \
+               --index-url https://download.pytorch.org/whl/nightly/rocm5.7;; \
```
Do you know whether base image rocm/pytorch:rocm6.1.2_ubuntu20.04_py3.9_pytorch_staging works for pytorch rocm5.7 and rocm 6.0 pytorch wheels?

mawong-amd replied:
Not sure if I understand this question, or the spirit behind the question: is the question whether "pytorch_staging" works on ROCm 5.7 or 6.0 without wheels, or whether these wheels work on a non-"pytorch_staging" image? Because you'd never install rocm5.7 or rocm6.0 wheels on a rocm6.1.2 image.

For the first question, the answer is "no": you need at least PyTorch 2.4.0 for `torch.cuda._device_count_amdsmi`, while `pytorch_staging` goes up to PyTorch 2.3.0. For the second, the answer is "yes": the relevant wheels for each ROCm version are designed to (and work on) any official base image from the relevant versions.

hongxiayang replied:
sorry for the confusing question. I was thinking about whether we should use the "staging" as the base image as the name means its testing status. And also whether rocm-6.0 + torch 2.4  or rocm-5.7 + torch 2.4, works out of box, for certain users on certain devices (especially, navi3x users, for example), because in the past the supported based docker images has torch 2.0.1 or torch 2.1.1.

mawong-amd replied:
"Staging" in this case refers to the PyTorch version (2.3.0, which we are targeting for stable release sometime), so it's a moot point when we're upgrading to a PyTorch 2.4.0 nightly. And this particular nightly has not raised problems on a wide range of internal testing on multiple GPUs architectures, including MI300, MI250, and Navi32.

Let's keep in mind the broader picture here: whether this PR is an improvement to the status quo, as opposed to whether this PR is "perfect". The status quo has multiple failing tests, even on the tests AMD supports, and most importantly, does not have ROCm 6.1 support, which brings in a host of performance improvements. And upstream vLLM has moved such that PyTorch 2.3+/2.4+ on ROCm is needed for a full complement of tests.

We can offer this support: even though on paper it is "experimental", it has not raised issues in testing. Unless someone provides a specific failing example that arises as a result of this PR, quibbling about potential issues that may arise because of the "experimental" or "nightly" status, without doing the legwork to verify that such issues actually exist, seems to me to be an example of making the perfect the enemy of the good.

mawong-amd commented:
> > > 1. Update PagedAttention hyperparameters
> 
> I agree about pulling out the change for "NUM_THREADS = 1024" to a different pr. As we discussed previously, do you know how this change impact non-mi300x uses?

This change will be pulled out of this PR, we can discuss the performance implications across devices in much more detail in that subsequent PR.

mawong-amd committed:
Remove sccache so it does not shadow ccache

mawong-amd committed:
Propagate CUDA_VISIBLE_DEVICES to HIP_VISIBLE_DEVICES if set before init

mawong-amd committed:
Revert "Add ROCm tuned num_threads and partition_size for pagedattn"

This reverts commit 75d5583ff83b5ef98ab97455ea0e7b5b8867331f.

mawong-amd committed:
Revert "Ensure use of docker buildkit for AMD container"

This reverts commit 834d006bdc3fa5ab1a5cb6d73bf55f44bd3de8d2.

mawong-amd committed:
Fix HIP_VISIBLE_DEVICE updating in device_count_stateless

mawong-amd commented:
Hi @WoosukKwon, the PR has now removed the PagedAttention tuning changes. I've addressed some of your feedback and answered the rest, please let me know if this PR is in an acceptable state! We would like to finally get ROCm 6.1 to users, as well as fix tests that have been broken.

The PR might touch many areas, which is not ideal, but this reflects more the fact that some upstream changes have subtly broken various tests on ROCm and not been caught. This is a bit of a chicken-and-egg situation: the longer our tests remain broken, the more likely broken PRs will be accepted. So we'd like to get these fixes in as soon as possible to forestall further breakage.

Thank you very much!

mawong-amd commented:
Now AMD CI is all green!

mawong-amd committed:
Robustify cuda_device_count_stateless logic on ROCm

mawong-amd committed:
Skip xfail tests on ROCm to conserve CI resources

WoosukKwon commented on Dockerfile.rocm:
```diff
@@ -39,79 +40,158 @@ RUN apt-get update && apt-get install -y \
+            patch /opt/rocm/include/hip/amd_detail/amd_hip_bf16.h rocm_patch/rocm_bf16.patch;; \
+        *"rocm-6.1"*) \
+            # Bring in upgrades to HIP graph earlier than ROCm 6.2 for vLLM
+            wget -N https://github.com/ROCm/vllm/raw/main/rocm_patch/libamdhip64.so.6 -P rocm_patch \
```
nit: What about using a specific commit version instead of `main` just for potential path changes or version changes?

mawong-amd replied:
Yup I was thinking in the off-chance we update it, it can be fetched automatically. But it's probably better to have a pin for reproducibility and update it if we ever change it, so I've pinned it to a recent commit.

WoosukKwon replied:
Thanks!

WoosukKwon commented on vllm/worker/worker_base.py:
```diff
@@ -125,6 +125,8 @@ def update_environment_variables(envs: Dict[str, str]) -> None:
             # overwriting CUDA_VISIBLE_DEVICES is desired behavior
             # suppress the warning in `update_environment_variables`
             del os.environ[key]
+            if is_hip():
+                os.environ.pop("HIP_VISIBLE_DEVICES", None)
```
Can we raise a warning if `HIP_VISIBLE_DEVICES` is not None?

mawong-amd replied:
Have added a warning and some flavor text to emphasize that if both `CUDA_VISIBLE_DEVICES` and `HIP_VISIBLE_DEVICES` is set, `CUDA_VISIBLE_DEVICES` takes precedence. Hopefully we can revert the need for this approach once ROCm-PyTorch starts recognizing `CUDA_VISIBLE_DEVICES` correctly.

WoosukKwon commented on vllm/distributed/device_communicators/custom_all_reduce_utils.py:
```diff
@@ -24,7 +25,8 @@ def producer(batch_src: Sequence[int],
              cuda_visible_devices: Optional[str] = None):
     if cuda_visible_devices is not None:
-        os.environ["CUDA_VISIBLE_DEVICES"] = cuda_visible_devices
+        update_environment_variables(
+            {"CUDA_VISIBLE_DEVICES": cuda_visible_devices})
```
@youkaichao We are considering exposing (only) `CUDA_VISIBLE_DEVICES` to AMD GPU users and convert it into `HIP_VISIBLE_DEVICES` internally in vLLM (inside `update_environment_variables`). Could you confirm that the code change here is safe? I also wanted to make sure you are aware of this change since this could be a bit error prone.

youkaichao replied:
makes sense to me, as long as the tests pass.

WoosukKwon replied:
Thanks @youkaichao!

WoosukKwon commented on tests/models/test_llava_next.py:
What about moving these test changes into a separate PR? We will then merge the PR almost immediately.

mawong-amd replied:
I've reverted test changes and will be putting it in a separate PR, but bear in mind that the AMD Vision Language Models test will most likely fail without these skips, because for some reason the CI machines take too long to fetch models.

We're still figuring this out—local runs pass on a machine which fetches the models quickly. For now it's of little consequence, because these models are xfail, but we'd like to have these run correctly eventually.

mawong-amd replied:
Tentatively seems to no longer be a problem! Good call on reverting these: no longer needed. vLLM moves so quickly 😊

WoosukKwon replied:
Great! Glad to hear that :)

mawong-amd committed:
Address reviewer comments

Revert "Skip xfail tests on ROCm to conserve CI resources"

This reverts commit 01fa95f7862ea52b19d96c16a5e1f7752cff3577.

DarkLight1337 commented:
> @DarkLight1337 It seems `runtime_env={"working_dir": VLLM_PATH}` was introduced in your PR #5100. Could you confirm that this change is safe and explain the reason if not?

It's unsafe to remove. Otherwise, some vLLM modules may become unavailable to ray workers during testing.

mawong-amd commented:
> > @DarkLight1337 It seems `runtime_env={"working_dir": VLLM_PATH}` was introduced in your PR #5100. Could you confirm that this change is safe and explain the reason if not?
> 
> It's unsafe to remove. Otherwise, some vLLM modules may become unavailable to ray workers during testing.

Can you point out where this happens? Tests were working fine before your PR, and I haven't yet seen any errors from removing it.

DarkLight1337 commented:
> > > @DarkLight1337 It seems `runtime_env={"working_dir": VLLM_PATH}` was introduced in your PR #5100. Could you confirm that this change is safe and explain the reason if not?
> > 
> > 
> > It's unsafe to remove. Otherwise, some vLLM modules may become unavailable to ray workers during testing.
> 
> Can you point out where this happens? Tests were working fine before your PR, and I haven't yet seen any errors from removing it.

I'm not an expert in `ray`, but I believe that it happened because `ray.init` is called inside test directory rather than the main vLLM package directory. Not sure why this problem doesn't happen now. I guess you can remove it as long as it doesn't break Ci.

mawong-amd committed:
Add amdsmi support for GPU memory check

mawong-amd commented:
@WoosukKwon Rebased and fixed a couple more conflicts, including adding AMD support for GPU memory clear (`amdsmi` in place of `nvml`) to fix distributed tests, now we have an all-green CI again!

WoosukKwon reviewed (approved):
@mawong-amd LGTM! This is big. Huge thanks for your effort!

WoosukKwon commented:
@hongxiayang While I understand your concerns mentioned in this PR, I will merge the PR as it includes some critical bug fixes and improvements. However, please feel free to open a new PR to continue the discussion.

hongxiayang commented:
> @hongxiayang While I understand your concerns mentioned in this PR, I will merge the PR as it includes some critical bug fixes and improvements. However, please feel free to open a new PR to continue the discussion.

Thanks @WoosukKwon for reviewing the PR and merge it.
